src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -237,7 +237,15 @@ private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
             var workspaceRegistration = (WorkspaceRegistration)sender;
             uint docCookie;
-            Contract.ThrowIfFalse(_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie));
+
+            // external workspace that listens to events from IVSRunningDocumentTable4 (just like MiscellaneousFilesWorkspace does) and registers the text buffer for opened document prior to Miscellaneous workspace getting notified of the docCookie - the prior contract assumes that MiscellaneousFilesWorkspace is always the first one to get notified
+            if (!_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie))
+            {
+                // We haven't even started tracking the document corresponding to this registration - likely because some external workspace registered the document's buffer prior to MiscellaneousWorkspace getting notified of it.
```

Comment:
Investigating further, this is a race condition in MiscellaneousWorkspace. Assume there is another external workspace that listens to IVSRunningDocumentTable4 events and as soon as text buffer is initialized it invokes OnDocumentOpened on one of the event callbacks for IVSRunningDocumentTable4. This will invoke [Workspace.RegisterText](http://source.roslyn.io/Microsoft.CodeAnalysis.Workspaces/R/5f1b56e57bf93158.html), which kicks off a schedule task to raise workspace changed events. Now assume this task runs simulatenously with [this](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,226) line in MiscellaneousFilesWorkspace.TrackOpenedDocument. MiscellaneousFilesWorkspace has already registered the WorkspaceChanged handler but not updated the _docCookieToWorkspaceRegistration map yet.
Given that this race can happen only if an external workspace has claimed the document, we would never attach to document [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,230) and hence [this](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,263) will be true, but [this](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,266) will be false and we would anyways do nothing in the handler. We still might want to put more thought into this overall scenario and other possible races.